libcxx/include/__config

```diff
@@ -1048,6 +1048,10 @@ __sanitizer_verify_double_ended_contiguous_container(const void*, const void*, c
 #    define _LIBCPP_HAS_NO_RTTI
 #  endif
 
+#  if !defined(_LIBCPP_HAS_NO_EXCEPTIONS) && !defined(_LIBCPP_HAS_NO_RTTI) && !defined(__APPLE__) && !defined(_WIN32)
```

Comment:
Why is this disabled on Apple and on Windows? It seems to me like this should be an availability macro instead so that all platforms can take advantage of it as long as the deployment target is recent enough to support this new addition. I think this is what @philnik777 was referring to in another comment.

Reply:
The "availability macro": is that https://clang.llvm.org/docs/AttributeReference.html#availability?
Anyway, i'm failing to understand how that could eliminate the need to have a runtime check for the `__cxa_init_primary_exception` presence -- say, if i build against 18.0 libcxxabi and then LD_PRELOAD 17.0 version, which should be a supported case, right?

-------

To answer your question directly:
Windows: disabled due to mingw requiring this https://github.com/llvm/llvm-project/blob/ea1909f82ceca30b009701f1d349247f8fbf1c3e/libcxxabi/include/__cxxabi_config.h#L101 to be specified for exception destructor, and this attribute now has to appear in libc++ at the call site (where weak declaration of the `__cxa_init_primary_exception` is present). I didn't want to copy-paste this macro in libc++, but definitely doable, should i implement it?

Apple: back-deployment-macosx tests fail (https://buildkite.com/llvm-project/libcxx-ci/builds/30969#018b50e0-c82b-42e0-bceb-b74d5808ffad, for example) with `dyld[47810]: missing symbol called` error. I have very limited knowledge about how linking is performed on Apple, but my quick googling tells that it's much different from what ld does, and weak attributes don't work without some compiler flags specified. I have no clue how this should be fixed and it also takes forever for CI to reach this check, so any help would be appreciated.

Reply:
I've added a mingw support

Reply:
If you look at how the availability macros in `__availability` work, basically we use the `--target <arch>-<vendor>-<platform><version>` to drive this. The idea is that you tell the compiler what's the oldest version of the platform you are targeting, and based on that we make the function available or not. If you then decide to run your program on an older platform, you effectively lied to the compiler and in practice you'll get a load-time error.